pkg/process/config.go

```diff
@@ -0,0 +1,223 @@
+			}
+			fullKey := base + key
+
+			// since this key can't affect anything from the config file and must be present
+			// on the command line, remove it so as to not mislead anyone
+			if fullKey == "defaults" {
+				continue
+			}
+
+			// gather information about the flag under consideration
+			var (
+				changed    bool
+				setup      bool
+				hidden     bool
+				user       bool
+				deprecated bool
+				comment    string
+				typ        string
+
+				_, overrideExists = options.Overrides[fullKey]
+			)
+			if f := flags.Lookup(fullKey); f != nil { // first check pflags
+				changed = f.Changed
+				setup = readBoolAnnotation(f, "setup")
+				hidden = readBoolAnnotation(f, "hidden")
+				user = readBoolAnnotation(f, "user")
+				deprecated = readBoolAnnotation(f, "deprecated")
+				comment = f.Usage
+				typ = f.Value.Type()
+			} else if f := flag.Lookup(fullKey); f != nil { // then stdlib flags
+				changed = f.Value.String() != f.DefValue
+				comment = f.Usage
+			} else {
+				// by default we store config values we know nothing about. we
+				// absue the meaning of "changed" to include this case.
+				changed = true
+			}
+
+			// in any of these cases, don't store the key in the config file
+			if setup || hidden || options.RemoveDeprecated && deprecated {
+				continue
+			}
+
+			// viper is super cool and doesn't cast floats automatically, so we
+			// handle that ourselves.
+			if typ == "float64" {
+				value = cast.ToFloat64(value)
+			}
+
+			flatKeys = append(flatKeys, fullKey)
+			flat[fullKey] = configValue{
+				value:   value,
+				comment: comment,
+				set:     user || changed || overrideExists,
+			}
+		}
+	}
+	filterAndFlatten("", settings)
+	sort.Strings(flatKeys)
+
+	// step 3. write out the configuration file
+	//
+
+	var nl = []byte("\n")
+	var lines [][]byte
+	for _, key := range flatKeys {
+		config := flat[key]
+
+		if config.comment != "" {
+			lines = append(lines, []byte("# "+config.comment))
+		}
+
+		data, err := yaml.Marshal(map[string]interface{}{key: config.value})
+		if err != nil {
+			return errs.Wrap(err)
+		}
+		dataLines := bytes.Split(bytes.TrimSpace(data), nl)
+
+		// if the config value is set, concat in the yaml lines
+		if config.set {
+			lines = append(lines, dataLines...)
+		} else {
+			// otherwise, add them in but commented out
+			for _, line := range dataLines {
+				lines = append(lines, append([]byte("# "), line...))
+			}
+		}
+
+		// add a blank line separator
+		lines = append(lines, nil)
+	}
+
+	return errs.Wrap(atomicWrite(outfile, 0600, bytes.Join(lines, nl)))
+}
+
+// readBoolAnnotation is a helper to see if a boolean annotation is set to true on the flag.
+func readBoolAnnotation(flag *pflag.Flag, key string) bool {
+	annotation := flag.Annotations[key]
+	return len(annotation) > 0 && annotation[0] == "true"
+}
+
+// atomicWrite is a helper to atomically write the data to the outfile.
+func atomicWrite(outfile string, mode os.FileMode, data []byte) (err error) {
```

Comment:
does this work right on windows? do we need atomic writes?

Reply:
yes, assuming it’s not replacing a directory.

Reply:
i think that's the Linux rules? on phone but Linux you can't rename over a directory iirc, and windows you need to remove the target first. what do we gain over just opening the destination directly for this?

Reply:
i went over this with andrew when we were pairing and we agreed it would work fine on windows. here’s the relevant code from the stdlib.  https://golang.org/src/internal/syscall/windows/syscall_windows.go#L259

indeed, our other code that does renames does the same thing except with `MOVEFILE_WRITE_THROUGH` (https://github.com/storj/storj/blob/master/storage/filestore/dir_windows.go#L124) which as per the docs at https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-movefileexa should only matter if the move is performed as a copy, which this one cannot. 

in general, i prefer atomic file writing, especially for important values like secrets that may not be easily replicated.

Reply:
i am placated